lib/http_aws_sigv4.c

```diff
@@ -66,6 +68,186 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
+{
+  for(;
+       *iptr;
+       iptr++) {
+    if(in_query && *iptr == '&') {
+      ++iptr;
+      goto out;
+    }
+    if(*iptr == '%' && iptr[1] && iptr[2]) {
+      char num_str[3] = {iptr[1], iptr[2], 0};
+      long ascii_code;
+
+      ascii_code = strtoul(num_str, NULL, 16);
+      if(ascii_code) {
+        /* this long is never bigger than 255 anyway */
+        *optr++ = curlx_ultouc(ascii_code);
+        iptr += 2;
+        continue;
+      }
+    }
+    *optr++=*iptr;
+  }
+out:
+  *optr = 0;
+  return iptr;
+}
+
+/* encode_query_cpy() copies a url to a output buffer and URL-encodes
+ * the URI acordingly
+ * this fucntion is a modified version of strcpy_url in urlapi.c
+ */
+static void encode_query_cpy(char *optr, const char *iptr, bool in_query)
+{
+  /* we must add this with whitespace-replacing */
+  int count_equal = 0;
+
+  for(;
+      *iptr;         /* until zero byte */
+      iptr++) {
+
+    switch(*iptr) {
+    case '=':
+      if(!in_query || count_equal++) {
+        msnprintf(optr, 4, "%%%02X", *iptr);
+        optr += 3;
+      }
+      else
+        *optr++=*iptr;
+      break;
+    default:
+      if(urlchar_needs_escaping(*iptr, in_query)) {
+        msnprintf(optr, 4, "%%%02X", (const unsigned char)*iptr);
+        optr += 3;
+      }
+      else
+        *optr++=*iptr;
+      break;
+    }
+  }
+  if(in_query && !count_equal)
+    *optr++= '=';
+  *optr = 0; /* null-terminate output buffer */
+}
+
+static char *new_encoded_url(char **src, bool in_query)
+{
+  char *ret = NULL;
+  char *decoded_tmp = NULL;
+  size_t canonical_len = encode_query_len(*src, in_query);
+  char *tmp;
+
+  if(!canonical_len)
+    return NULL;
+  /* the size is not exact, but at worst, bigger of a few bytes */
+  decoded_tmp = malloc(canonical_len + 1);
+  if(!decoded_tmp)
+    return NULL;
+  tmp = decode_query_cpy(decoded_tmp, *src, in_query);
+  if(in_query && tmp)
+    *src = tmp;
+
+  ret = malloc(canonical_len + 1);
+  if(ret)
+    encode_query_cpy(ret, decoded_tmp, in_query);
+  free(decoded_tmp);
+  return ret;
+}
+
+static char *new_encoded_query_url(char *src)
+{
+  struct curl_slist *query_list = NULL;
+  char *reenc = NULL;
+  char *ret = NULL;
+  struct curl_slist *l;
+  int again;
+
+  while((reenc = new_encoded_url(&src, TRUE)) != NULL) {
+    l = Curl_slist_append_nodup(query_list, reenc);
+    if(!l) {
+      free(reenc);
+      goto out;
+    }
+    query_list = l;
+  }
+
+  /* reorder */
+  for(again = 1; again;) {
+    again = 0;
+    for(l = query_list; l; l = l->next) {
+      struct curl_slist *next = l->next;
+
+      if(next && strcmp(l->data, next->data) > 0) {
+        char *tmp = l->data;
+
+        l->data = next->data;
+        next->data = tmp;
+        again = 1;
+      }
+    }
+  }
+
+  for(l = query_list; l; l = l->next) {
+    char *tmp = ret;
+
+    ret = curl_maprintf("%s%s%s", ret ? ret : "", ret ? "&": "", l->data);
+    free(tmp);
+  }
```

Comment:
Can I propose that you build a `dynbuf` string here instead? It'll be *much* more efficient. Something like:
~~~c
struct dynbuf buff;

Curl_dyn_init(&buff, CURL_MAX_QUERY_LENGTH);

for(l = query_list; l; l = l->next) {
  if(Curl_dyn_len(&buff))
    if(Curl_dyn_add(&buff, "&", 1))
      goto error;
  if(Curl_dyn_add(&buff, l->data))
    goto error;
}

return Curl_dyn_ptr(&buff);